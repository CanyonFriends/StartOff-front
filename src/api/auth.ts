import axios from 'axios';
import { SignupInfoType } from '../validator/signupValidator';
import { LoginInfoType } from '../validator/loginValidator';
import { ErrorType } from './error';

export interface SigninResponseType {
  access_token: string;
  uuid: string;
  email: string;
  nickname: string;
  user_id: string;
}

export interface SignupResponseType {
  user_id: string;
}

export interface LogoutRequestType {
  email: string;
  accessToken: string;
  uuid: string;
}

export interface LogoutResponseType {
  msg: string;
  success: boolean;
}

export const signupAPI = async ({ id, pw, nickname }: SignupInfoType): Promise<ErrorType | SignupResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/signup',
      data: {
        nickname,
        email: id,
        password: pw,
      },
    });
    return response.data as SignupResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const signinAPI = async ({ id, pw }: LoginInfoType): Promise<ErrorType | SigninResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/login',
      data: {
        email: id,
        password: pw,
      },
    });
    return response.data as SigninResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const logoutAPI = async ({
  email,
  accessToken,
  uuid,
}: LogoutRequestType): Promise<ErrorType | LogoutResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/logout',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      data: {
        email,
        accessToken,
        uuid,
      },
    });
    return response.data as LogoutResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};
