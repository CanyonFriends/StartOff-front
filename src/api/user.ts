import axios from 'axios';
import { SignupInfoType } from '../validator/signupValidator';
import { LoginInfoType } from '../validator/loginValidator';
import { ErrorType } from './error';

export interface SigninResponseType {
  access_token: string;
  refresh_token: string;
  email: string;
  token_type: string;
  user_id: string;
}

export const signupAPI = async ({ id, pw, nickname }: SignupInfoType) => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/signup',
      data: {
        nickname,
        email: id,
        password: pw,
      },
    });
    return response;
  } catch (error) {
    return { error: error.response.data.errorMsg } as ErrorType;
  }
};

export const signinAPI = async ({ id, pw }: LoginInfoType): Promise<ErrorType | SigninResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/login',
      data: {
        email: id,
        password: pw,
      },
    });
    return response.data as SigninResponseType;
  } catch (error) {
    return { error: error.response.data.errorMsg } as ErrorType;
  }
};
