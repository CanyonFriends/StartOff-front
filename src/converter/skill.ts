import { SkillClientType } from '../@types/client';
import { SkillServerType } from '../@types/server';

export const skillServerType2ClientType = (skill: SkillServerType): SkillClientType => {
  return {
    skillId: String(skill.skill_id),
    color: skill.color,
    skillName: skill.skill_name,
    textColor: skill.text_color,
  };
};
