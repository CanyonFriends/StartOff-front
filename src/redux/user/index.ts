import { createSlice } from '@reduxjs/toolkit';
import * as reducers from './reducer';
import { UserState } from './types';

export const initialState: UserState = {
  userId: '',
  nickname: '',
  email: '',
  error: '',
  isSignin: false,
};

export const userSlice = createSlice({
  initialState,
  name: 'user',
  reducers,
});

export const userReducer = userSlice.reducer;
export const { actions } = userSlice;
