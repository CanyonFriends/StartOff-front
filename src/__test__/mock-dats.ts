import { ErrorType } from '../api/error';
import { SigninResponseType, SignupResponseType } from '../api/user';

export const signinSuccessMockInfo: SigninResponseType = {
  access_token: 'access_token',
  refresh_token: 'refresh_token',
  email: 'email@email.com',
  token_type: 'bearer',
  user_id: '1',
};

export const signinFailMockInfo: ErrorType = {
  errorMsg: '에러 발생',
};

export const signupSuccessMockInfo: SignupResponseType = {
  user_id: '1',
};

export const signupFailMockInfo: ErrorType = {
  errorMsg: '에러 발생',
};
