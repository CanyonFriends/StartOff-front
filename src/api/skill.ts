import axios from '../utils/axios';
import { ErrorType } from './error';
import { SkillServerType } from '../@types/server';
import { skillServerType2ClientType } from '../converter/skill';
import { SkillType } from '../@types/client';

export const getSkillsAPI = async (): Promise<SkillType[] | ErrorType> => {
  try {
    const response = await axios({
      method: 'GET',
      url: `/v1/skills`,
    });

    return (response.data as SkillServerType[]).map((skill) => skillServerType2ClientType(skill));
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};
