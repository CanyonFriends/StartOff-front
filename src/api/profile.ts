import { ProjectServerResponseType, SkillServerType } from '../@types/server';
import axios from '../utils/axios';
import { ErrorType } from './error';
import { ProfileType, SkillType } from '../@types/client';
import { profileResponse2Type } from '../converter/profile';
import { skillServerType2ClientType } from '../converter/skill';

export interface GetProfileRequest {
  userId: string;
}

export interface GetProfileResponse {
  baekjoon_id: string;
  blog_url: string;
  github_url: string;
  introduce: string;
  nickname: string;
  projects: ProjectServerResponseType[];
  user_skills: SkillServerType[];
}

export interface UpdateProfileIntroduceRequest {
  userId: string;
  introduce: string;
  nickname: string;
}

export interface UpdateProfileGithubRequest {
  userId: string;
  githubUrl: string;
}

export interface UpdateProfileBlogRequest {
  userId: string;
  blogUrl: string;
}

export interface UpdateUserSkillRequest {
  userId: string;
  skillName: string;
}

export interface DeleteUserSkillRequest {
  userId: string;
  skillId: string;
}

export const getProfileAPI = async ({ userId }: GetProfileRequest): Promise<ProfileType | ErrorType> => {
  try {
    const response = await axios({
      method: 'GET',
      url: `/v1/users/${userId}/profile`,
    });
    return profileResponse2Type(response.data as GetProfileResponse);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const updateProfileIntroduce = async ({
  userId,
  introduce,
  nickname,
}: UpdateProfileIntroduceRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/introduce`,
      data: {
        introduce,
        nickname,
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

// TODO: 이름 변경
export const updateGithubIntroduce = async ({
  userId,
  githubUrl,
}: UpdateProfileGithubRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/github-url`,
      data: {
        github_url: githubUrl,
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const updateBlogIntroduce = async ({
  userId,
  blogUrl,
}: UpdateProfileBlogRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/blog-url`,
      data: {
        blog_url: blogUrl,
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const updateUserSkillAPI = async ({
  userId,
  skillName,
}: UpdateUserSkillRequest): Promise<SkillType | ErrorType> => {
  try {
    const response = await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/skills`,
      data: {
        skill_name: skillName,
      },
    });
    return skillServerType2ClientType(response.data as SkillServerType);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const deleteUserSkillAPI = async ({ userId, skillId }: DeleteUserSkillRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'DELETE',
      url: `/v1/users/${userId}/skills/${skillId}`,
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};
