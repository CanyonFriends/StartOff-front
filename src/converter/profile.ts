import { ProfileType, ProjectType, SkillType } from '../@types/client';
import { ProjectServerType, SkillServerType } from '../@types/server';
import { GetProfileResponse } from '../api/profile';

export const skillServerType2ClientType = (skill: SkillServerType): SkillType => {
  return {
    skillId: String(skill.skill_id),
    color: skill.color,
    skillName: skill.skill_name,
    textColor: skill.text_color,
  };
};

export const projectServerType2ClientType = (project: ProjectServerType): ProjectType => {
  return {
    content: project.content,
    deployUrl: project.deploy_url,
    endDate: project.end_date,
    githubUrl: project.github_url,
    id: project.id,
    introduce: project.introduce,
    projectSklls: project.project_skills.map((skill) => skillServerType2ClientType(skill)),
  };
};

export const profileResponse2Type = (profileResponse: GetProfileResponse): ProfileType => {
  return {
    baekjoonId: profileResponse.baekjoon_id,
    blogUrl: profileResponse.blog_url,
    githubUrl: profileResponse.github_url,
    introduce: profileResponse.introduce,
    nickname: profileResponse.nickname,
    projects: profileResponse.projects.map((project) => projectServerType2ClientType(project)),
    userSkills: profileResponse.user_skills.map((skill) => skillServerType2ClientType(skill)),
  };
};
