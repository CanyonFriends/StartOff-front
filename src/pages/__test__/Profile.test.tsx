/**
 * @jest-environment jsdom
 */
import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react';
import { render } from '../../test-utils';
import ProfilePage from '../Profile';
import { getProfileAPI, updateProfileIntroduce, updateGithubIntroduce, updateBlogIntroduce } from '../../api/profile';
import { getSkillsAPI } from '../../api/skill';
import { profile, skills, project } from '../../__test__/mock-dats';
import { dateToString } from '../../utils/date';
import { updatePasswordAPI } from '../../api/user';

jest.mock('../../api/user');
jest.mock('../../api/profile');
jest.mock('../../api/skill');

const updatePassworMociAPI = updatePasswordAPI as jest.MockedFunction<typeof updatePasswordAPI>;
const getProfileMockAPI = getProfileAPI as jest.MockedFunction<typeof getProfileAPI>;
const getSkillsMockAPI = getSkillsAPI as jest.MockedFunction<typeof getSkillsAPI>;
const updateProfileIntroduceMockAPI = updateProfileIntroduce as jest.MockedFunction<typeof updateProfileIntroduce>;
const updateGithubIntroduceMock = updateGithubIntroduce as jest.MockedFunction<typeof updateGithubIntroduce>;
const updateBlogIntroduceMock = updateBlogIntroduce as jest.MockedFunction<typeof updateBlogIntroduce>;

beforeEach(() => {
  jest.resetAllMocks();
  getProfileMockAPI.mockReturnValue(new Promise((res) => res(profile)));
  getSkillsMockAPI.mockReturnValue(new Promise((res) => res(skills)));
});

describe('<Profile> 페이지', () => {
  it('snapshot 체크', async () => {
    const component = render(<ProfilePage />);
    await waitFor(() => expect(component.container).toMatchSnapshot());
  });

  describe('비밀번호 변경 기능', () => {
    it('비밀번호 변경 모달 렌더링', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modalButton = component.getByText('계정 정보 변경');
        fireEvent.click(modalButton);

        const currentPasswordInput = component.getByLabelText('currentpw') as HTMLInputElement;
        const afterPasswirdInput = component.getByLabelText('afterpw') as HTMLInputElement;
        const confirmPasswordinput = component.getByLabelText('confirmpw') as HTMLInputElement;
        expect(currentPasswordInput.value).toBe('');
        expect(afterPasswirdInput.value).toBe('');
        expect(confirmPasswordinput.value).toBe('');
      });
    });

    it('비밀번호 변경', async () => {
      updatePassworMociAPI.mockReturnValue(new Promise((res) => res(true)));
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modalButton = component.getByText('계정 정보 변경');
        fireEvent.click(modalButton);

        const currentPasswordInput = component.getByLabelText('currentpw');
        const afterPasswirdInput = component.getByLabelText('afterpw');
        const confirmPasswordinput = component.getByLabelText('confirmpw');

        fireEvent.change(currentPasswordInput, { target: { value: `currentpassword` } });
        fireEvent.change(afterPasswirdInput, { target: { value: `afterpassword` } });
        fireEvent.change(confirmPasswordinput, { target: { value: `confirmpassword` } });
      });
    });
  });

  describe('introduce기능', () => {
    it('렌더링', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        component.getByText(profile.nickname);
        component.getByText(profile.introduce);
      });
    });

    it('수정버튼 클릭 시 전환', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modifyButton = component.getByText('수정');
        fireEvent.click(modifyButton);

        // input박스 전환여부 확인
        const nicknameInput = component.getByLabelText('nickname') as HTMLInputElement;
        const introduceInput = component.getByLabelText('introduce') as HTMLInputElement;
        expect(nicknameInput.value).toBe(profile.nickname);
        expect(introduceInput.value).toBe(profile.introduce);
      });
    });

    it('수정 전환 후 저장 전환', async () => {
      updateProfileIntroduceMockAPI.mockReturnValue(new Promise((res) => res(true)));
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modifyButton = component.getByText('수정');
        fireEvent.click(modifyButton);
        const nicknameInput = component.getByLabelText('nickname') as HTMLInputElement;
        const introduceInput = component.getByLabelText('introduce') as HTMLInputElement;
        fireEvent.change(nicknameInput, { target: { value: `${profile.nickname}!` } });
        fireEvent.change(introduceInput, { target: { value: `${profile.introduce}!` } });

        const saveButton = component.getByText('저장');
        fireEvent.click(saveButton);
        component.getByText(`${profile.nickname}!`);
        component.getByText(`${profile.introduce}!`);
      });
    });
  });

  describe('github url기능', () => {
    it('렌더링', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        component.getByText(profile.githubUrl);
      });
    });

    it('수정 아이콘 클릭 시 전환', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modifyButton = component.getByLabelText('logo-pencil');
        fireEvent.click(modifyButton);

        const githubInput = component.getByLabelText('github') as HTMLInputElement;
        expect(githubInput.value).toBe(profile.githubUrl);
      });
    });

    it('수정 전환 후 저장 전환', async () => {
      updateGithubIntroduceMock.mockReturnValue(new Promise((res) => res(true)));
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modifyButton = component.getByLabelText('logo-pencil');
        fireEvent.click(modifyButton);
        const githubInput = component.getByLabelText('github') as HTMLInputElement;
        fireEvent.change(githubInput, { target: { value: `${profile.githubUrl}!` } });

        const saveButton = component.getByLabelText('logo-disk');
        fireEvent.click(saveButton);
        component.getByText(`${profile.githubUrl}!`);
      });
    });
  });

  describe('blog url기능', () => {
    it('렌더링', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        component.getByText(profile.blogUrl);
      });
    });

    it('수정 아이콘 클릭 시 전환', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modifyButton = component.getByLabelText('home-pencil');
        fireEvent.click(modifyButton);

        const blogInput = component.getByLabelText('blog') as HTMLInputElement;
        expect(blogInput.value).toBe(profile.blogUrl);
      });
    });

    it('수정 전환 후 저장 전환', async () => {
      updateBlogIntroduceMock.mockReturnValue(new Promise((res) => res(true)));
      const component = render(<ProfilePage />);

      await waitFor(() => {
        const modifyButton = component.getByLabelText('home-pencil');
        fireEvent.click(modifyButton);
        const blogInput = component.getByLabelText('blog') as HTMLInputElement;
        fireEvent.change(blogInput, { target: { value: `${profile.blogUrl}!` } });

        const saveButton = component.getByLabelText('home-disk');
        fireEvent.click(saveButton);
        component.getByText(`${profile.blogUrl}!`);
      });
    });
  });

  describe('기술스택 기능', () => {
    it('렌더링', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        component.getByText(profile.userSkills[0].skillName);
        component.getByText(profile.userSkills[1].skillName);
      });
    });
  });

  describe('프로젝트 기능', () => {
    it('렌더링', async () => {
      const component = render(<ProfilePage />);

      await waitFor(() => {
        component.getByText(project.title);
        component.getByText(project.deployUrl);
        component.getByText(project.githubUrl);
        component.getByText(project.introduce);
        component.getByText(`${dateToString(project.startDate)} ~ ${dateToString(project.endDate)}`);
        component.getByText(project.content);
        component.getByText(skills[0].skillName);
      });
    });
  });
});
