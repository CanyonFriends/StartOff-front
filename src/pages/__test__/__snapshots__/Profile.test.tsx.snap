// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile> 페이지 snapshot 체크 1`] = `
<div>
  <div
    id="modal"
  />
</div>
`;
